### ISSUE TITLE
Inserttags do not parse within the breadcrumb module

### ISSUE THREAD

**Issue by zoglo** [contao/contao#7105]
### Affected version(s)

5.3.x

### Description

It was common to use the insert tag `{{page::pageTitle}}` for detail pages in news to generate the page title within the breadcrumb.

#### Static title
![image](https://github.com/contao/contao/assets/55794780/92d79963-e11a-4d06-82bc-20b52462d2d0)

This seems broken in Contao 5.3.x and can even be seen in the demo when using `{{page::pageTitle}}`:
*Edit* 
~~https://demo.contao.org/news/new-contao-versions-in-short-intervalls.html~~ (See screenshot)

#### `{{page::pageTitle}}`
![image](https://github.com/contao/contao/assets/55794780/1bdefe69-b0a0-4176-8db9-7b9f747386be)

**Comment by fritzmg** [comment]
Sort of a duplicate of https://github.com/contao/contao/issues/6050. While the cause in #6050 is different (here it is caused by the more granular replacement of insert tags) the eventual solution likely needs to be the same.

// edit: oh, they are not replaced with the page title at all? That's something different then may be.

**Comment by asaage** [comment]
it works if the breadcrumb comes after the reader
https://community.contao.org/de/showthread.php?86441-order-of-operations&p=581453#post581453
I'd appreciate if somebody finds a solution where you don't even have to insert ```{{page::pageTitle}}``` into the pageName because it makes the PageTreeview in the Backend look like a hack.

**Comment by zoglo** [comment]
> I'd appreciate if somebody finds a solution where you don't even have to insert {{page::pageTitle}} into the pageName because it makes the PageTreeview in the Backend look like a hack.


@asaage feel free to adjust the logic and implement it for Contao. It ain't flawless and needs some tests as well:
https://github.com/oveleon/contao-glossary-bundle/blob/main/src/EventListener/BreadcrumbListener.php

That would most likely be a feature tho

**Comment by zoglo** [comment]
After debugging, this happens most likely within these lines:
https://github.com/contao/contao/blob/2c84decd5a81e16a508186060733bdf5deeb43be/core-bundle/contao/library/Contao/InsertTags.php#L107-L120

since it outputs $tags with a size of 1, thus returning a ChunkedText.

**Comment by ausi** [comment]
> After debugging, this happens most likely within these lines:

what do you mean by that?

**Comment by zoglo** [comment]
> what do you mean by that?

Basically how far my investigation went.
$tags is actually an array that looks like this:
```php
$tags = [
    '',
    '{{page::pageTitle}},
    ''
]
```
Basically, the regex does not work because the expected output should look like this:
```php
$tags = [
    'page',
    'pageTitle',
]
```

**Comment by ausi** [comment]
> $tags is actually an array that looks like this:

No, it looks like `['', 'page::pageTitle', '']` _(the middle part without `{{` and `}}`)_ 
But anyway, this is not related to this issue as far as I can tell, as the `{{page::pageTitle}}` insert tag itself works just fine, only in the breadcrumb module there seems to be an issue, right?

**Comment by zoglo** [comment]
Yeah, debugged late at night so maybe a wrong conclusion

**Comment by netzarbeiter** [comment]
As an addition: If you add a headline with the content `{{page::pageTitle}}` to the news detail page the output in Contao 5 is "News and updates" the page title of the page. In Contao 4.13 it was the title of the news "New Contao Versions in short intervalls".

https://demo.contao.org/preview.php/news/new-contao-versions-in-short-intervalls.html

**Comment by asaage** [comment]
> As an addition: If you add a headline with the content {{page::pageTitle}} to the news detail page

I noticed the same while using a detail-page with the reader-module from plenta/contao-jobs-basic-bundle.
Is it possible to delay the resolving of this Inserttag ```{{page::pageTitle}}``` (probably others as well) until every module that changes something has done it's thing? The culprit seems to be located there more than actualy in the breadcrumb module.

**Comment by fritzmg** [comment]
I still think that the correct solution for this issue are - in my opinion - reader pages.

**Comment by frontendschlampe** [comment]
I find this issue after having same problems updating a website to Contao 5. Is there any solution for the near future, because that's a showstopper for any update. Every client and their SEO experts will complain about this.

The issue is nearly one year old …

**Comment by MacKP** [comment]
The workaround with the breadcrumb underneath the reader is not really an useful option. In that case i need JS for moving the breadcrumb in the dom back to the top (pure css is not a solution because the logik strukture would be differnt).

Kind regards

**Comment by fritzmg** [comment]
> The issue is nearly one year old …

If there was an easy solution, somebody would have implemented it already ;). Either someone from the core team - or someone else from the community.

**Comment by frontendschlampe** [comment]
I haven't made any statements about feasibility, because I can't assess that at all. I just wanted to say that this feature is essential for any website that uses reader pages in conjunction with breadcrumb navigation. And it's not as if it's a new feature, it worked up to version 5 with a template adjustment. And as @MacKP has already written: it is absolutely problematic in combination with accessibility.

**Comment by fritzmg** [comment]
You also made a comment about how old this issue is, to which I responded.

**Comment by ausi** [comment]
> this feature is essential for any website that uses reader pages in conjunction with breadcrumb navigation.

Agree. This needs to be fixed.

> the correct solution for this issue are - in my opinion - reader pages.

I agree as well ☺️

Until we have the correct solution I want to try two things:

1. Render the reader module ~~after~~ _before_ the other modules and elements. This might be hard/impossible to do especially if they are nested inside articles (which is the most common case AFAIK).

2. If 1. does not work, render all `{{page::*}}` insert tags lazily and only replace them when rendering the fe_page template. However this is not a nice solution (string replace on the whole HTML page is bad and wrong) and might break other things.

**Comment by fritzmg** [comment]
> 1. Render the reader module after the other modules and elements.

You mean the breadcrumb module, right? If the breadcrumb renders after the reader module, everything would be fine.

**Comment by leofeyer** [comment]
I agree that replacing the `{{page::pageTitle}}` insert tag is too intrusive. In my opinion, it should work like this (simplified):

```php
$breadcrumbNonce = null;
$breadcrumbModule = null;

foreach ($modules as $module) {
    if ($module->isBreadcrumb()) {
        $breadcrumbNonce = 'unique string';
        $breadcrumbModule = $module;
        $return .= $breadcrumbNonce;
    } else {
        $return .= $module->generate();
    }
}

if ($breadcrumbNonce) {
    $return = str_replace($breadcrumbNonce, $breadcrumbModule->generate(), $return);
}
```

This would delay generating the breadcrumb module. Of course, it will not work if the module is added via insert tag or if the reader module is added via insert tag, but it should fix the default case.

**Comment by ausi** [comment]
> You mean the breadcrumb module, right?

I meant rendering the reader module before (not after 🙈). I updated my comment.

**Comment by ausi** [comment]
> In my opinion, it should work like this (simplified):

Agree, this could be one of the possible implementations for (1.), but I’d try to implement it without string replace first ☺️
